### PR TITLE
Support opset-11 Gemm kernels

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Gemm now supports optional bias in its spec. 

**Motivation and Context**
Make ORT opset-11 compliant

### CODE REVIEW

File: onnxruntime/core/providers/cpu/math/gemm_helper.h [hunk 1]
```diff
@@ -50,17 +50,17 @@ class GemmHelper {
   Status State() const { return status_; }
 
  private:
-  bool IsValidBroadcast(const TensorShape& shape, int64_t M, int64_t N) {
-    if (shape.NumDimensions() != 1 && shape.NumDimensions() != 2)
+  bool IsValidBroadcast(const TensorShape& bias_shape, int64_t M, int64_t N) {
+    if (bias_shape.NumDimensions() != 0 && bias_shape.NumDimensions() != 1 && bias_shape.NumDimensions() != 2)
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/math/gemm_helper.h]: should support scalar bias as well (i.e.) bias_shape.NumDimensions() == 0

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc [hunk 2]
```diff
@@ -347,7 +347,7 @@ class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Un
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Det);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, ScatterElements);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, NonMaxSuppression);
-
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kOnnxDomain, 11, Gemm);
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/cpu_execution_provider.cc]: There's some overlap with [https://github.com/microsoft/onnxruntime/pull/1956](https://github.com/microsoft/onnxruntime/pull/1956) and may need an update to the opset 9 declaration if that gets merged first.

File: onnxruntime/core/providers/cpu/math/gemm.cc [hunk 3]
```diff
@@ -11,4 +11,10 @@ ONNX_CPU_OPERATOR_VERSIONED_KERNEL(
     9,
     KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
     Gemm<float>);
-}
+
+ONNX_CPU_OPERATOR_KERNEL(
+    Gemm,
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/math/gemm.cc]: Would be good to have a comment saying bias was made optional

File: onnxruntime/core/providers/cuda/math/gemm.cc [hunk 4]
```diff
@@ -38,25 +48,26 @@ template <typename T>
 Status Gemm<T>::ComputeInternal(OpKernelContext* ctx) const {
   typedef typename ToCudaType<T>::MappedType CudaT;
 
-  const auto X = ctx->Input<Tensor>(0);
-  const auto W = ctx->Input<Tensor>(1);
-  const auto B = ctx->Input<Tensor>(2);
-  GemmHelper helper(X->Shape(), trans_A_, W->Shape(), trans_B_, B->Shape());
+  const auto* X = ctx->Input<Tensor>(0);
+  const auto* W = ctx->Input<Tensor>(1);
+  const auto* B = ctx->InputCount() == 3 ? ctx->Input<Tensor>(2) : nullptr;
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/math/gemm.cc]: >ctx->InputCount() == 3 ? [](start = 18, length = 24)

ctx->Input does a check on input count so it's not really necessary here

File: onnxruntime/core/providers/cpu/math/gemm_helper.h (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/math/gemm_helper.h]: Could this just be bias_shape.NumDimension() > 2

---
In reply to: [328366840](https://github.com/microsoft/onnxruntime/pull/1923#discussion_r328366840) [](ancestors = 328366840)

File: onnxruntime/core/providers/cuda/math/gemm.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cuda/math/gemm.cc]: I see. So I presume it returns a nullptr if the input doesn't exist ?

File: onnxruntime/core/providers/cpu/math/gemm.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/math/gemm.cc]: sure.

File: onnxruntime/core/providers/cpu/math/gemm_helper.h (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/math/gemm_helper.h]: Sure. Will change this.

File: onnxruntime/core/providers/cuda/math/gemm.cc (hunk 4 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cuda/math/gemm.cc]: Yes

File: onnxruntime/core/providers/cpu/math/gemm.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/math/gemm.cc]: Added a comment.

File: onnxruntime/core/providers/cuda/math/gemm.cc (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cuda/math/gemm.cc]: Made the change

File: onnxruntime/core/providers/cpu/cpu_execution_provider.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/cpu_execution_provider.cc]: Fixed.

**Pull request merged**